### DISCUSSION TITLE
Functions to compute OID

### DISCUSSION THREAD

**Discussion by jeffparsons** [GitoxideLabs/gitoxide#151] [Ideas]
I have this code in a project:

```rust
trait TreeExt {
    fn compute_oid(&self) -> ObjectId;
}

impl TreeExt for mutable::Tree {
    // Based on <https://git-scm.com/book/en/v2/Git-Internals-Git-Objects>.
    fn compute_oid(&self) -> ObjectId {
        let mut buf: Vec<u8> = Vec::new();
        self.write_to(&mut buf)
            .expect("Failed to write; shouldn't happen because we were writing to a Vec");
        let header = format!("tree {}\0", buf.len());
        let mut sha1 = sha1::Sha1::new();
        sha1.update(header.as_bytes());
        sha1.update(&buf);
        let sha1_digest = sha1.digest();
        ObjectId::from_borrowed_sha1(&sha1_digest.bytes())
    }
}
```

Would you be open to a PR that introduces an equivalent (but non-allocating) function to all relevant types in git-object? Seems like it might be generally useful.

**Comment by Byron** [comment]
I see how that could be useful.

In case you are looking for a place to put such a trait, `git-repository` [has a place for that](https://github.com/Byron/gitoxide/blob/f8e07475f8867fc98a9264b1270977b48283a94e/git-repository/src/ext.rs#L1-L67).

There is [an in-memory object store](https://github.com/Byron/gitoxide/blob/2958145a0ae1ef582bbf88352f5567d5c2b5eaf0/git-odb/src/store/sink.rs#L12) already which is should be useful for this kind of functionality.
